### PR TITLE
core/parsigex: add verification function to parsigex

### PR DESCRIPTION
**Pull request by dB2510**
Adds a signature verification in parsigex along with Verification function constructors: `NewEth2Verifier` for core workflow and  `newDKGVerifier` for dkg/exchanger.

category: feature
ticket: #217

### CODE REVIEW

File: core/parsigex/parsigex.go [hunk 1]
```diff
@@ -50,11 +56,12 @@ func NewParSigEx(tcpNode host.Host, sendFunc p2p.SendFunc, peerIdx int, peers []
...
-	tcpNode  host.Host
-	sendFunc p2p.SendFunc
-	peerIdx  int
-	peers    []peer.ID
-	subs     []func(context.Context, core.Duty, core.ParSignedDataSet) error
+	tcpNode    host.Host
+	sendFunc   p2p.SendFunc
+	peerIdx    int
+	peers      []peer.ID
+	verifyFunc func(context.Context, core.PubKey, core.Duty, core.ParSignedData) error
```
Comment (corverroos) [inline core/parsigex/parsigex.go:63]: nit: swap duty and pubkey, as that is the standard order we use in the code base

File: core/parsigex/parsigex.go [hunk 2]
```diff
@@ -89,6 +96,14 @@ func (m *ParSigEx) handle(s network.Stream) {
 	ctx, span = core.StartDutyTrace(ctx, duty, "core/parsigex.Handle")
 	defer span.End()
 
+	// Verify partial signature
+	for pubkey, data := range set {
+		if err = m.verifyFunc(ctx, pubkey, duty, data); err != nil {
+			log.Error(ctx, "Invalid signature from peer", err, z.Str("peer", p2p.PeerName(s.Conn().RemotePeer())))
```
Comment (corverroos) [inline core/parsigex/parsigex.go:102]: `Peer exchanged invalid partial signature`

File: core/parsigex/parsigex.go [hunk 3]
```diff
@@ -130,3 +145,60 @@ func (m *ParSigEx) Broadcast(ctx context.Context, duty core.Duty, set core.ParSi
 func (m *ParSigEx) Subscribe(fn func(context.Context, core.Duty, core.ParSignedDataSet) error) {
 	m.subs = append(m.subs, fn)
 }
+
+// NewEth2Verifier returns a verifyFunc instance which is responsible to verify the given partial signatures.
+func NewEth2Verifier(eth2Svc eth2client.Service, pubSharesByKey map[core.PubKey][]*bls_sig.PublicKey) func(context.Context, core.PubKey, core.Duty, core.ParSignedData) error {
```
Comment (corverroos) [inline core/parsigex/parsigex.go:150]: Put this `core` package

Comment (dB2510) [reply core/parsigex/parsigex.go:150]: putting this in core package will introduce cyclic dependency

Comment (dB2510) [reply core/parsigex/parsigex.go:150]: core => signing => tblsconv => core

File: core/parsigex/parsigex.go [hunk 4]
```diff
@@ -130,3 +145,60 @@ func (m *ParSigEx) Broadcast(ctx context.Context, duty core.Duty, set core.ParSi
 func (m *ParSigEx) Subscribe(fn func(context.Context, core.Duty, core.ParSignedDataSet) error) {
 	m.subs = append(m.subs, fn)
 }
+
+// NewEth2Verifier returns a verifyFunc instance which is responsible to verify the given partial signatures.
+func NewEth2Verifier(eth2Svc eth2client.Service, pubSharesByKey map[core.PubKey][]*bls_sig.PublicKey) func(context.Context, core.Duty, core.PubKey, core.ParSignedData) error {
+	eth2Cl := eth2Svc.(signing.Eth2Provider)
```
Comment (corverroos) [inline core/parsigex/parsigex.go:151]: never cast without checking for errors

File: core/parsigex/parsigex.go [hunk 5]
```diff
@@ -130,3 +145,60 @@ func (m *ParSigEx) Broadcast(ctx context.Context, duty core.Duty, set core.ParSi
 func (m *ParSigEx) Subscribe(fn func(context.Context, core.Duty, core.ParSignedDataSet) error) {
 	m.subs = append(m.subs, fn)
 }
+
+// NewEth2Verifier returns a verifyFunc instance which is responsible to verify the given partial signatures.
+func NewEth2Verifier(eth2Svc eth2client.Service, pubSharesByKey map[core.PubKey][]*bls_sig.PublicKey) func(context.Context, core.Duty, core.PubKey, core.ParSignedData) error {
```
Comment (corverroos) [inline core/parsigex/parsigex.go:150]: move this to core package

File: core/parsigex/parsigex_test.go [hunk 6]
```diff
@@ -77,7 +84,9 @@ func TestParSigEx(t *testing.T) {
 	// create ParSigEx components for each host
 	for i := 0; i < n; i++ {
 		wg.Add(n - 1)
-		sigex := parsigex.NewParSigEx(hosts[i], p2p.Send, i, peers)
+		sigex := parsigex.NewParSigEx(hosts[i], p2p.Send, i, peers, func(_ context.Context, _ core.Duty, _ core.PubKey, _ core.ParSignedData) error {
```
Comment (corverroos) [inline core/parsigex/parsigex_test.go:87]: can remove all `_`

File: core/validatorapi/validatorapi.go [hunk 7]
```diff
@@ -703,45 +706,39 @@ func (c Component) getProposerPubkey(ctx context.Context, duty core.Duty) (core.
...
-
-// verifySignedData aliases signing.VerifySignedData skipping it for insecureTest.
-func (c Component) verifySignedData(ctx context.Context, eth2Cl signing.Eth2Provider,
-	pubkey core.PubKey, eth2SignedType interface{},
-) error {
-	if c.insecureTest {
-		return nil
+	// eth2util shouldn't import core package, so can't use core.SignedData.
+	// To avoid pointer vs non-pointer issues, always get value if pointer is provided.
+	if reflect.TypeOf(eth2SignedType).Kind() == reflect.Pointer {
```
Comment (corverroos) [inline core/validatorapi/validatorapi.go:724]: loosing type safety isn't worth it. Suggest:
```
// Call this indirection function that will call the callback if !insecureTest and with the pubshare if found. Then no need to `interface{}` or reflection.
err := c.verifyPartialSig(pubkey, func(pubshare ) error {
   return signing.VerifyBlock(ctx, c.eth2Cl, pubshare, signed)
}
if err != nil {
  return nil
}
```

**Review by corverroos — CHANGES_REQUESTED** [review]
This PR is too big, split the validaotrapi stuff from the DKG stuff please

**Pull request closed without merging**